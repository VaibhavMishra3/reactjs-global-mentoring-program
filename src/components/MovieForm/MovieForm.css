.movie-form {
  color: white;
  background-color: #232323;
  padding: 20px;
}

.form-group {
  margin-bottom: 15px;
}

.form-group label {
  display: block;
  margin-bottom: 5px;
}

.form-group input,
.form-group textarea {
  width: 100%;
  padding: 10px;
  background-color: #424242;
  border: 1px solid #555555;
  border-radius: 5px;
  color: white;
}

.genre-select {
  display: flex;
  flex-wrap: wrap;
  gap: 10px;
}

.genre-option {
  display: flex;
  align-items: center;
}

.genre-option input {
  margin-right: 5px;
}

.error-message {
  color: #f65261;
  font-size: 12px;
}

.form-buttons {
  display: flex;
  justify-content: space-between;
  margin-top: 20px;
}

.reset-button,
.submit-button {
  padding: 10px 20px;
  border: none;
  border-radius: 5px;
  cursor: pointer;
}

.reset-button {
  background-color: #555555;
}

.submit-button {
  background-color: #f65261;
  color: white;
}

.submit-button:disabled {
  background-color: #555555;
}
