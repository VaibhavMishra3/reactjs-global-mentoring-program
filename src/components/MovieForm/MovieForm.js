import React from "react";

function MovieForm({ initialMovie = {}, onSubmit }) {
  const handleSubmit = (e) => {
    e.preventDefault();
    const formData = Object.fromEntries(new FormData(e.target));
    onSubmit(formData);
  };

  return (
    <form onSubmit={handleSubmit}>
      <div>
        <label>Title</label>
        <input
          type="text"
          name="title"
          defaultValue={initialMovie.title || ""}
          required
        />
      </div>
      <div>
        <label>Release Date</label>
        <input
          type="date"
          name="releaseDate"
          defaultValue={initialMovie.releaseDate || ""}
          required
        />
      </div>
      <div>
        <label>Movie URL</label>
        <input
          type="url"
          name="movieUrl"
          defaultValue={initialMovie.movieUrl || ""}
          required
        />
      </div>
      <div>
        <label>Rating</label>
        <input
          type="number"
          name="rating"
          min="0"
          max="10"
          step="0.1"
          defaultValue={initialMovie.rating || ""}
          required
        />
      </div>
      <div>
        <label>Genre</label>
        <select name="genre" defaultValue={initialMovie.genre || ""} required>
          <option value="">Select Genre</option>
          <option value="Action">Action</option>
          <option value="Comedy">Comedy</option>
          <option value="Drama">Drama</option>
        </select>
      </div>
      <div>
        <label>Runtime</label>
        <input
          type="text"
          name="runtime"
          defaultValue={initialMovie.runtime || ""}
          required
        />
      </div>
      <div>
        <label>Overview</label>
        <textarea
          name="overview"
          defaultValue={initialMovie.overview || ""}
        ></textarea>
      </div>
      <div>
        <button type="reset">Reset</button>
        <button type="submit">Submit</button>
      </div>
    </form>
  );
}

export default MovieForm;
