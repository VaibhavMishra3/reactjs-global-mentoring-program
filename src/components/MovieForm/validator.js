import { isNotEmpty, isValidDate, isValidURL } from '../../utils/validators.js';

export const INITIAL_VALUES = {
  'title': '',
  'release_date': '',
  'poster_path': '',
  'vote_average': '',
  'genres': [],
  'runtime': '',
  'overview': ''
};

export const validator = values => {
  const errors = {};

  for (const property in values) {
    if (values[property] === INITIAL_VALUES[property]) {
      errors[property] = 'This field is required';
    }
  }

  if (isNotEmpty(values.vote_average) && (values.vote_average < 0 || values.vote_average > 10)) {
    errors.vote_average = 'This field must be from 0 to 10 inclusive';
  }

  if (isNotEmpty(values.release_date) && !isValidDate(values.release_date)) {
    errors.release_date = 'This field must be a YYYY-MM-DD date';
  }

  if (isNotEmpty(values.poster_path) && !isValidURL(values.poster_path)) {
    errors.poster_path = 'This field must be a valid URL';
  }

  if (isNotEmpty(values.runtime) && values.runtime <= 0) {
    errors.runtime = 'This field must be greater than 0';
  }

  return errors;
};
