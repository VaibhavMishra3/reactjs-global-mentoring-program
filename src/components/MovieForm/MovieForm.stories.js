import MovieForm from './MovieForm';
import { MOCK_MOVIES } from '../../constants/mockData.js';

export default {
  component: MovieForm
};

export const Default = {
  args: {
    movie: MOCK_MOVIES[0],
    onSubmit: (event) => console.log('form submitted: ' + JSON.stringify(event))
  },
  parameters: {
  }
};
