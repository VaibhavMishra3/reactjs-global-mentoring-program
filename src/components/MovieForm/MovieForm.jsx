import React from 'react';
import { PropTypes } from 'prop-types';
import { Formik, Form } from 'formik';
import { Container, Row, Modal, Button } from 'react-bootstrap';

import AreaField from '../AreaField/AreaField';
import SelectField from '../SelectField/SelectField';
import TextField from '../TextField/TextField';
import { GENRES } from '../../constants/data.js';
import { movieType } from '../../constants/types.js';
import { INITIAL_VALUES, validator } from './validator.js';

const MovieForm = ({ movie, onSubmit }) => {
  return (
    <Formik
      initialValues={movie || INITIAL_VALUES}
      validate={validator}
      onSubmit={(values, { setSubmitting }) => {
        onSubmit(values);
      }}
    >
      {({
        isSubmitting
      }) => (
        <Form data-testid={`${movie ? 'edit-movie-form' : 'add-movie-form'}`} noValidate>
          <Modal.Body>
            <Container>
              <Row>
                <TextField
                  type='text'
                  label='Title'
                  name='title'
                  placeholder='Enter title'
                />
                <TextField
                  type='text'
                  label='Release date'
                  name='release_date'
                  placeholder='Enter release date'
                />
              </Row>
              <Row>
                <TextField
                  type='text'
                  label='Movie URL'
                  name='poster_path'
                  placeholder='https://'
                />
                <TextField
                  type='number'
                  label='Rating'
                  name='vote_average'
                  placeholder='7.8'
                />
              </Row>
              <Row>
                <SelectField
                  name='genres'
                  label='Genre'
                  options={GENRES}
                />
                <TextField
                  type='number'
                  label='Runtime'
                  name='runtime'
                  placeholder='minutes'
                />
              </Row>
              <Row>
                <AreaField
                  label='Overview'
                  name='overview'
                  placeholder='Movie description'
                />
              </Row>
            </Container>
          </Modal.Body>
          <Modal.Footer>
            <Button variant='secondary' type='reset'>
              Reset
            </Button>
            <Button variant='primary' type='submit' disabled={isSubmitting}>
              Submit
            </Button>
          </Modal.Footer>
        </Form>
      )}
    </Formik>
  );
}

MovieForm.propTypes = {
  movie: movieType,
  onSubmit: PropTypes.func.isRequired
};

export default MovieForm;
