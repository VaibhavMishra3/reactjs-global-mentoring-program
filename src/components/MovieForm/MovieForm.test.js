import React from 'react';
import renderer from 'react-test-renderer';
import { cleanup, render, screen, act, fireEvent } from '@testing-library/react'

import MovieForm from './MovieForm';
import { MOCK_MOVIES } from '../../constants/mockData.js';

describe('MovieForm', () => {

  afterEach(cleanup);

  it('should match snapshot', () => {
    const movieForm = renderer
      .create(
        <MovieForm
          movie={MOCK_MOVIES[0]}
          onSubmit={console.log}
        />
      )
      .toJSON();

    expect(movieForm).toMatchSnapshot();
  });

  it('should call the submit handler', async () => {
    const handleSubmitMock = jest.fn();

    render(
      <MovieForm
        movie={MOCK_MOVIES[0]}
        onSubmit={handleSubmitMock}
      />
    );

    const submitButton = screen.getAllByRole('button')[1];
    expect(submitButton).toHaveAttribute('type', 'submit');
    expect(submitButton.className).toBe('btn btn-primary');

    await act(async () => {
      await fireEvent.click(submitButton, {});
    });

    expect(handleSubmitMock).toHaveBeenCalled();
  });

});
