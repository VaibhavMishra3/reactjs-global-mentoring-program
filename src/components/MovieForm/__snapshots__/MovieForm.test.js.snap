// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieForm should match snapshot 1`] = `
<form
  action="#"
  data-testid="edit-movie-form"
  noValidate={true}
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="modal-body"
  >
    <div
      className="container"
    >
      <div
        className="row"
      >
        <div
          className="col"
        >
          <label
            className="form-label"
          >
            Title
          </label>
          <input
            className="form-control"
            name="title"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Enter title"
            type="text"
            value="Pulp Fiction"
          />
        </div>
        <div
          className="col"
        >
          <label
            className="form-label"
          >
            Release date
          </label>
          <input
            className="form-control"
            name="release_date"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Enter release date"
            type="text"
            value="1994-09-10"
          />
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col"
        >
          <label
            className="form-label"
          >
            Movie URL
          </label>
          <input
            className="form-control"
            name="poster_path"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="https://"
            type="text"
            value="https://image.tmdb.org/t/p/w500/dM2w364MScsjFf8pfMbaWUcWrR.jpg"
          />
        </div>
        <div
          className="col"
        >
          <label
            className="form-label"
          >
            Rating
          </label>
          <input
            className="form-control"
            name="vote_average"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="7.8"
            type="number"
            value={8.3}
          />
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col"
        >
          <label
            className="form-label"
          >
            Genre
          </label>
          <select
            className="form-control"
            multiple={true}
            name="genres"
            onBlur={[Function]}
            onChange={[Function]}
            value={
              Array [
                "Thriller",
                "Crime",
              ]
            }
          >
            <option
              value="Action"
            >
              Action
            </option>
            <option
              value="Adventure"
            >
              Adventure
            </option>
            <option
              value="Comedy"
            >
              Comedy
            </option>
            <option
              value="Crime"
            >
              Crime
            </option>
            <option
              value="Fantasy"
            >
              Fantasy
            </option>
            <option
              value="History"
            >
              History
            </option>
            <option
              value="Science Fiction"
            >
              Science Fiction
            </option>
            <option
              value="Thriller"
            >
              Thriller
            </option>
            <option
              value="Western"
            >
              Western
            </option>
          </select>
        </div>
        <div
          className="col"
        >
          <label
            className="form-label"
          >
            Runtime
          </label>
          <input
            className="form-control"
            name="runtime"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="minutes"
            type="number"
            value={154}
          />
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col"
        >
          <label
            className="form-label"
          >
            Overview
          </label>
          <textarea
            className="form-control"
            name="overview"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Movie description"
            value="A burger-loving hit man, his philosophical partner, a drug-addled gangster's moll and a washed-up boxer converge in this sprawling, comedic crime caper. Their adventures unfurl in three stories that ingeniously trip back and forth in time."
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="modal-footer"
  >
    <button
      className="btn btn-secondary"
      disabled={false}
      type="reset"
    >
      Reset
    </button>
    <button
      className="btn btn-primary"
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
